mlflow/cli.py

```diff
@@ -145,49 +146,86 @@ def run(uri, entry_point, version, param_list, experiment_name, experiment_id, m
-              help="Additional command line options forwarded to gunicorn processes.")
-def ui(backend_store_uri, default_artifact_root, host, port, gunicorn_opts):
-    """
-    Launch the MLflow tracking UI.
+def _server_options(function):
+    function = click.option("--backend-store-uri", "--file-store", metavar="PATH",
+                            default=DEFAULT_LOCAL_FILE_AND_ARTIFACT_PATH,
+                            help="URI or path for backend store implementation. Acceptable backend"
+                                 "store are SQLAlchemy compatible implementation or local storage."
+                                 " Supports various SQLAlchemy compatible database like SQLite,"
+                                 " MySQL, PostgreSQL. As an example MySQL backed store can be"
+                                 " configured using connection string."
+                                 "'mysql://<user_name>:<password>@<host>:<port>/<database_name>' "
+                                 "By default file based backed store"
+                                 " will be used. (default: ./mlruns).")(function)
+    function = click.option("--default-artifact-root", metavar="URI", default=None,
+                            help="Local or S3 URI to store artifacts, for new experiments. "
+                                 "Note that this flag does not impact already-created experiments. "
+                                 "Default: " + DEFAULT_LOCAL_FILE_AND_ARTIFACT_PATH)(function)
+    function = click.option("--host", "-h", metavar="HOST", default="127.0.0.1",
+                            help="The network address to listen on (default: 127.0.0.1). "
+                                 "Use 0.0.0.0 to bind to all addresses if you want to"
+                                 " access the tracking server from other machines.")(function)
+    function = click.option("--port", "-p", default=5000,
+                            help="The port to listen on (default: 5000).")(function)
+    function = click.option("--gunicorn-opts", default=None,
+                            help="Additional command line options forwarded to gunicorn"
+                                 " processes.")(function)
+    return function
 
-    The UI will be visible at http://localhost:5000 by default.
-    """
-    # Ensure that both backend_store_uri and default_artifact_uri are set correctly.
+
+def _run_tracking_server(backend_store_uri, default_artifact_root, host, port, workers,
+                         static_prefix, gunicorn_opts):
     if not backend_store_uri:
         backend_store_uri = DEFAULT_LOCAL_FILE_AND_ARTIFACT_PATH
 
     if not default_artifact_root:
         if _is_local_uri(backend_store_uri):
             default_artifact_root = backend_store_uri
         else:
-            default_artifact_root = DEFAULT_LOCAL_FILE_AND_ARTIFACT_PATH
+            eprint("Option 'default-artifact-root' is required, when backend store is not "
```

Comment:
Here's the context for this choice of defaults, from #860 (cc: @aarondav)

1. `mlflow ui` : is a mechanism to connect to an existing backend store for the purpose of accessing existing MLflow data. Like for instance connecting to a remote SQL store. Also users can use this to try put MLflow. For this reason we default both `backend_store_uri`, `default_artifact_root` to a local folder `./mlruns`. When using a specialized backend store like a SQL store, if `default_artifact_root` is not provided, we see what the `backend_store_uri` is -- if local folder, we use that to store artifacts else we revert back to the default `./mlruns`.

2. `mlflow server` : is a more of a production of remote tracking server running on EC2 or k8s or something. In such cases, the users expectation is that the storage be more permanent. Hence we mandate that the user set an explicit backend and artifact store like a mounted NFS store etc. More details @ https://mlflow.org/docs/latest/tracking.html#storage.

@stbof : do you think we should document some of this?

Reply:
I think the distinction makes sense, I ended up changing it because only 1 branch in the execution path was different.

If `mflow server` is for prod. servers, wouldn't it make more sense to require the user to explicitly define the backend store URI? Otherwise, we face the same issue where the tracking server is reverting to defaults behind the scene.